source/MaterialXTest/ShaderGen.cpp

```diff
@@ -40,6 +41,98 @@ bool isTopologicalOrder(const std::vector<const mx::SgNode*>& nodeOrder)
     return true;
 }
 
+// Observer to add library markers to all nodes. Use the "namespace" attribute to do this
+static std::string NAMESPACE_ATTRIBUTE("namespace");
+class LibraryObserver : public MaterialX::Observer
+{
+public:
+    LibraryObserver() {}
+
+    void onAddElement(MaterialX::ElementPtr /*parent*/, MaterialX::ElementPtr element) override
+    {
+        if (_libraryRefName.length())
+        {
+            element->setAttribute(NAMESPACE_ATTRIBUTE, _libraryRefName);
+        }
+    }
+
+    void setLibraryRefName(const std::string& name)
+    {
+        _libraryRefName = name;
+    }
+
+protected:
+    // Libraryref name to add
+    std::string _libraryRefName;
+};
+
+//
```

Comment:
Hey Niklas, I have some of this functionality in the utility library and have copied over versions here.
Let me know if you think any of it is worth putting into ShaderGen. Thanks.

Reply:
Maybe we should wait until we settle for a solution on the namespace handling, as we're discussing this with ILM now.

Reply:
Sorry, the comment was for any form of the utilities below, not for the namespace push above. Yes, we should wait for this to be resolved.